### ISSUE TITLE
Immediate tokens and words

### ISSUE THREAD

**Issue by Caagr98** [tree-sitter/tree-sitter#706]
I'm trying to define strings for my language, by doing
```js
	string: $ => choice(
        seq('"', repeat(choice($.strchr, $._stresc)), token.immediate('"')),
        seq("'", repeat(choice($.strchr, $._stresc)), token.immediate("'")),
    ),
	strchr: $ => token.immediate(/[^\x00-\x1F\x7F-\x9F]/), // No unprintables
	_stresc: $ => seq(token.immediate("\\"), choice(
		alias(token.immediate("n"),  $.stresc_0A),
		alias(token.immediate("\\"),  $.stresc_5C),
        // other escapes
	)),

	name: $ => /[a-zA-Z_][a-zA-Z_0-9]*/, // word = $ => $.name
```

This works very well overall, since literal strings have precedence over `strchr`. However, it tries to match `name` tokens pretty much everywhere, meaning that while `"asdf\n"`works, `"\nasdf"` finds the `name` "nasdf" rather than the character "n". This obviously results in a parse error, despite being valid syntax.

This can be worked around by matching `\n` as one token, so that `, but then I need to add a `seq(token.immediate("\\"), token.immediate("\0"))` fallback to make unknown escape sequences give an error rather than being lexed into `strchar`, and it's just an overall dissatisfying workaround. This also isn't the only place where this problem happens; refactoring the `e` in float literals is much harder. I could also disable `word` entirely, but that has its own bunch of issues.

Since immediate tokens ignore `extras`, I find it quite strange that `name` isn't similarly ignored.

**Comment by maxbrunsfeld** [comment]
The issue body is a bit difficult to read because of markdown formatting problems. Maybe double check it.

Have you checked out how this is done in the existing tree-sitter grammars?

**Comment by Caagr98** [comment]
Oops, fixed formatting, sorry.

It seems most existing grammars just wrap the entire escape sequence rule in a `token()`. That works, obviously, but it doesn't give a very nice syntax tree.

**Comment by maxbrunsfeld** [comment]
> That works, obviously, but it doesn't give a very nice syntax tree.

I think creating a separate token for each individual character in a string isn't the best syntax tree representation. Because of all the internal nodes, it's going to be slower, both during parsing and when *using* the resulting syntax tree.

> I need to add a `seq(token.immediate("\\"), token.immediate("\0"))` fallback to make unknown escape sequences give an error rather than being lexed into strchar.

In general, I would recommend avoiding the trying to perform detailed *semantic* validation as part of the lexing/parsing process. Things like rules for valid escape sequences can be handled more efficiently and precisely (e.g. much better error messages) by a separate layer of hand-written code that inspects the syntax tree. I think it's usually better to err on the side of keeping the actual grammar simpler and 'looser', and enforce more specific invariants elsewhere.

**Comment by maxbrunsfeld** [comment]
> refactoring the `e` in float literals is much harder

Hmm, that's interesting. Do you want to give more detail on this?

Again, my recommendation would probably be to include the `e` in the float literal token, rather than parsing it as a separate node. I believe that is how most programming language lexers/parsers would do it.

**Comment by Caagr98** [comment]
Having a separate node per character seems to already be how most existing grammars ([json](https://github.com/tree-sitter/tree-sitter-json/blob/d3976b27df8622ed17bef6dd5e358b398e73c676/grammar.js#L39-L52), [python](https://github.com/tree-sitter/tree-sitter-python/blob/d00b1d4016bdaf74e24a1fbddcfd07e08deebcd0/grammar.js#L782-L820), [c](https://github.com/tree-sitter/tree-sitter-c/blob/99151b1e9293c9e025498fee7e6691e1a52e1d03/grammar.js#L917-L935)) already do it, so I don't see how that would be a problem — unless you're referring to my wrapping them in a named node rather than an anonymous one?

I agree that *semantic* analysis should generally be left to other tools, but I think escapes fall very much under syntax. My language contains some more complicated escape sequences, as well as raw strings, which when put together would make second-pass escape handling a somewhat tedious exercise. I would much prefer to just get that stuff directly in the syntax tree if possible.

I do agree that floats are probably better of treated as a single token, so they aren't a very good example. If one wanted to treat them as a nonterminal anyway, that would be something like
```js
float: $ => seq(
    $._float_without_exponent,
    optional(seq(
        token.immediate("e"),
        field("exponent", token.immediate(/[+-]?\d+/)),
    ),
),
```
This would make `1.2e4` be parsed as a `$._float_without_exponent` followed by the `$.name` `"e4"`. Again, not the best example; what I was trying to say was pretty much that while string escapes can generally be refactored into a form that doesn't interfere with `word`, that might not always be the case.

**Comment by maxbrunsfeld** [comment]
Yeah, I see your point about `word` causing problems if you have an `e` immediate token in a float. To be honest, I think I may have encountered a similar problem in one of my own grammars, when enabling `word`, but just fixed the problem by refactoring the rule and moved on.

I think you're right that this should be considered a bug in the behavior of `word`. Would you be up for writing up a minimal self-contained grammar and test case that reproduces the problem? Something where `word` changes behavior in an unexpected way.

**Comment by Caagr98** [comment]
Here's a (mostly) minimal grammar.
```js
module.exports = grammar({
	name: "word",
	// word: $ => $.name,

	rules: {
		string: $ => seq(
			'"',
			repeat(choice(
				$.stresc,
				$.strchr,
			)),
			optional(token.immediate('"')),
		),

		strchr: $ => token.immediate(/./),

		stresc: $ => seq(
			token.immediate("\\"),
			token.immediate("n"),
		),

		name: $ => /[a-zA-Z_][a-zA-Z_0-9]*/,
	},
});
```

It parses the string `"\na"` as
```lisp
(string [0, 0] - [1, 0]
  (stresc [0, 1] - [0, 3])
  (strchr [0, 3] - [0, 4]))
```
if `word` is disabled, and
```
(string [0, 0] - [1, 0]
  (ERROR [0, 1] - [0, 4]
    (name [0, 2] - [0, 4])))
```
if it is enabled.

As an aside, I noticed that ERROR nodes often include one or more `strchr` tokens, presumably because they can be matched almost anywhere. Probably nothing that can be done about that.

**Comment by maxbrunsfeld** [comment]
> Having a separate node per character seems to already be how most existing grammars (json, python, c) already do it, so I don't see how that would be a problem — unless you're referring to my wrapping them in a named node rather than an anonymous one?

Just to be clear though, this is not how the existing grammars do it. All of these existing grammars have a token that matches a sequence of *one or more* characters. So unless there are escape sequences in the string, a string would always consist of three tokens: two `"` tokens, and one hidden token representing the content. I still would really recommend refactoring to avoid having all of these individual `strchr` nodes.
